File: Week_02/G20190343020286/[49]Group Anagrams.java
//给定一个字符串数组，将字母异位词组合在一起。字母异位词指字母相同，但排列不同的字符串。 
//
// 示例: 
//
// 输入: ["eat", "tea", "tan", "ate", "nat", "bat"],
//输出:
//[
//  ["ate","eat","tea"],
//  ["nat","tan"],
//  ["bat"]
//] 
//
// 说明： 
//
// 
// 所有输入均为小写字母。 
// 不考虑答案输出的顺序。 
// 
// Related Topics 哈希表 字符串



//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
    /**
     * 将每个字符串排序，排序后字符串作为key,List作为value存入map中
     * 最后返回map的values
     * @param strs
     * @return
     */
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String,List<String>> map = new HashMap<>();
        List<List<String>> res = new ArrayList<>();
        for (int i = 0; i < strs.length; i++) {
            char[] ch = strs[i].toCharArray();
            Arrays.sort(ch);
            String key = new String(ch);
            if(!map.containsKey(key)){
                map.put(key,new ArrayList<String>());
            }
            map.get(key).add(strs[i]);
        }
        return new ArrayList<>(map.values());
    }


    /**
     * 方法二：将每个字符串对应26数组上，
     * 转换为0#0#2#3的字符串作为key，实际的字符串链表为value
     * @param strs
     * @return
     */
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String,List<String>> map = new HashMap<>();
        List<List<String>> res = new ArrayList<>();
        for (int i = 0; i < strs.length; i++) {
            int[] arr = new int[26];
            for (char ch : strs[i].toCharArray()) {
                arr[ch - 'a']++;
            }
            String key = "";
            for (int j = 0; j < arr.length; j++) {
                key += arr[j] + "#";
            }
            if(!map.containsKey(key)){
                map.put(key,new ArrayList<>());
            }
            map.get(key).add(strs[i]);
        }
        return new ArrayList<>(map.values());
    }
}
//leetcode submit region end(Prohibit modification and deletion)

File: Week_02/G20190343020014/[49]Group Anagrams.py
#给定一个字符串数组，将字母异位词组合在一起。字母异位词指字母相同，但排列不同的字符串。 
#
# 示例: 
#
# 输入: ["eat", "tea", "tan", "ate", "nat", "bat"],
#输出:
#[
#  ["ate","eat","tea"],
#  ["nat","tan"],
#  ["bat"]
#] 
#
# 说明： 
#
# 
# 所有输入均为小写字母。 
# 不考虑答案输出的顺序。 
# 
# Related Topics 哈希表 字符串


from typing import List
#leetcode submit region begin(Prohibit modification and deletion)

import collections
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        result = collections.defaultdict(list)
        for s in strs:
            result[tuple(sorted(s))].append(s)
        return result.values()


# class Solution:
#     def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
#         ans = collections.defaultdict(list)
#         for s in strs:
#             count = [0] * 26
#             for c in s:
#                 count[ord(c) - ord('a')] += 1
#             ans[tuple(count)].append(s)
#         return ans.values()

#leetcode submit region end(Prohibit modification and deletion)
r = Solution().groupAnagrams(["eat","tea","tan","ate","nat","bat"])
print(r)